File: src/gaia/daemon/paths.py
# Copyright(C) 2025-2026 Advanced Micro Devices, Inc. All rights reserved.
# SPDX-License-Identifier: MIT
"""On-disk locations for the daemon's single-instance state (design §0.26).

Everything lives under ``~/.gaia/host/``. ``GAIA_DAEMON_HOME`` overrides the
directory (used by tests so a run never clobbers the user's real daemon, and so
concurrent tests stay isolated). The override is read on every call — not cached —
so a subprocess spawned with a different env resolves its own directory.
"""

from __future__ import annotations

import json
import os
import shutil
from pathlib import Path
from typing import Union

_ENV_HOME = "GAIA_DAEMON_HOME"


def host_dir() -> Path:
    """Directory holding instance.json, the start lock, and the daemon log."""
    override = os.environ.get(_ENV_HOME)
    if override:
        return Path(override)
    return Path.home() / ".gaia" / "host"


def instance_path() -> Path:
    """The single-instance registry file (pid + port + client token)."""
    return host_dir() / "instance.json"


def lock_path() -> Path:
    """Advisory lock serializing daemon start (so two callers spawn one daemon)."""
    return host_dir() / "instance.lock"


def log_path() -> Path:
    """Daemon stdout/stderr log (what ``gaia daemon logs`` tails)."""
    return host_dir() / "daemon.log"


def sidecars_ledger_path() -> Path:
    """The sidecar spawn ledger (pids the daemon must reap after a hard crash)."""
    return host_dir() / "sidecars.json"


def ensure_host_dir() -> Path:
    """Create ``host_dir()`` if missing and return it."""
    d = host_dir()
    d.mkdir(parents=True, exist_ok=True)
    return d


def atomic_write_json(path: Path, payload, mode: int = 0o600) -> None:
    """Atomically persist *payload* as JSON at *path*, default file mode 0600.

    Shared by ``instance.py`` and ``sidecars/ledger.py`` — one copy of a
    secrets-adjacent write routine, not two that drift. Writes a uniquely-named
    temp file in the same directory (so ``os.replace`` is a same-filesystem
    atomic rename), fsyncs it, then renames it over the target. The temp file
    is created ``O_EXCL`` at *mode* so it is never briefly world-readable.
    """
    path = Path(path)
    path.parent.mkdir(parents=True, exist_ok=True)
    tmp = path.parent / f".{path.name}.{os.getpid()}.tmp"
    # O_EXCL: a leftover temp from a prior crashed writer must not be reused.
    if tmp.exists():
        tmp.unlink()
    fd = os.open(str(tmp), os.O_WRONLY | os.O_CREAT | os.O_EXCL | os.O_TRUNC, mode)
    try:
        with os.fdopen(fd, "w", encoding="utf-8") as f:
            f.write(json.dumps(payload, indent=2))
            f.flush()
            os.fsync(f.fileno())
    except Exception:
        try:
            tmp.unlink()
        except OSError:
            pass
        raise
    os.replace(str(tmp), str(path))
    # os.replace preserves the temp's mode; re-assert defensively for platforms
    # where the source mode is not carried across the rename.
    try:
        os.chmod(str(path), mode)
    except OSError:
        pass


def atomic_copy_file(
    src: Union[str, Path], dst: Union[str, Path], mode: int = 0o600
) -> None:
    """Atomically copy *src* to *dst* — same temp-then-rename discipline as
    :func:`atomic_write_json`, but for arbitrary (binary) files.

    Used by the one-time state migration to relocate SQLite stores: the copy is
    written to a uniquely-named temp file in *dst*'s directory, fsynced, then
    ``os.replace``-d over the target, so a crash mid-copy never leaves a
    half-written (and therefore corrupt) database at *dst*. *src* is never
    modified — the migration is non-destructive by construction.
    """
    src = Path(src)
    dst = Path(dst)
    dst.parent.mkdir(parents=True, exist_ok=True)
    tmp = dst.parent / f".{dst.name}.{os.getpid()}.tmp"
    if tmp.exists():
        tmp.unlink()
    fd = os.open(str(tmp), os.O_WRONLY | os.O_CREAT | os.O_EXCL | os.O_TRUNC, mode)
    try:
        with open(src, "rb") as fsrc, os.fdopen(fd, "wb") as fdst:
            shutil.copyfileobj(fsrc, fdst)
            fdst.flush()
            os.fsync(fdst.fileno())
    except Exception:
        try:
            tmp.unlink()
        except OSError:
            pass
        raise
    os.replace(str(tmp), str(dst))
    try:
        os.chmod(str(dst), mode)
    except OSError:
        pass

File: tui/internal/daemon/paths.go
// Package daemon is the TUI's thin client for the GAIA daemon: single-instance
// discovery, the start-or-attach handshake, and authenticated calls against the
// daemon's HTTP control plane (`/daemon/v1/*`) and agent relay (`/v1/<agent>/*`).
//
// It is the Go counterpart of src/gaia/daemon/{paths,instance,client}.py and
// deliberately mirrors their trust rules: a recorded instance is trusted only
// when its pid is alive AND a token-authed status probe answers with the daemon's
// service id and the recorded pid — a freed port that some unrelated process
// grabbed after a crash is a real failure mode, not a theoretical one.
//
// Every failure returns an actionable error naming what failed, what to run, and
// which log to read. There is no silent fallback to a direct sidecar connection:
// the TUI only ever holds the DAEMON client token, never a sidecar bearer.
package daemon

import (
	"fmt"
	"os"
	"path/filepath"
)

// EnvHome overrides the daemon state directory (tests, and any non-default
// install). Read on every call, never cached — a subprocess spawned with a
// different env must resolve its own directory.
const EnvHome = "GAIA_DAEMON_HOME"

// HostDir returns the directory holding instance.json, the start lock, and the
// daemon log.
func HostDir() (string, error) {
	if override := os.Getenv(EnvHome); override != "" {
		return override, nil
	}
	home, err := os.UserHomeDir()
	if err != nil {
		return "", fmt.Errorf(
			"cannot resolve the home directory to locate ~/.gaia/host: %w — "+
				"set %s to the daemon state directory and retry", err, EnvHome)
	}
	return filepath.Join(home, ".gaia", "host"), nil
}

// InstancePath returns the single-instance registry file (pid + port + client token).
func InstancePath() (string, error) {
	dir, err := HostDir()
	if err != nil {
		return "", err
	}
	return filepath.Join(dir, "instance.json"), nil
}

// LockPath returns the advisory lock that serializes daemon start, so two
// concurrent callers spawn exactly one daemon.
func LockPath() (string, error) {
	dir, err := HostDir()
	if err != nil {
		return "", err
	}
	return filepath.Join(dir, "instance.lock"), nil
}

// LogPath returns the daemon stdout/stderr log (what `gaia daemon logs` tails).
func LogPath() (string, error) {
	dir, err := HostDir()
	if err != nil {
		return "", err
	}
	return filepath.Join(dir, "daemon.log"), nil
}

// logPathForMessage returns LogPath() for embedding in an error string, or a
// human placeholder if even the home directory is unresolvable.
func logPathForMessage() string {
	p, err := LogPath()
	if err != nil {
		return "~/.gaia/host/daemon.log"
	}
	return p
}

// ensureHostDir creates the host directory if missing and returns it.
func ensureHostDir() (string, error) {
	dir, err := HostDir()
	if err != nil {
		return "", err
	}
	if err := os.MkdirAll(dir, 0o700); err != nil {
		return "", fmt.Errorf("cannot create the daemon state directory %s: %w", dir, err)
	}
	return dir, nil
}
